server/applicationset/applicationset.go

```diff
@@ -0,0 +1,361 @@
+			newItems = append(newItems, *a)
+		}
+	}
+	if q.Name != "" {
+		newItems, err = argoutil.FilterAppSetsByName(newItems, q.Name)
+		if err != nil {
+			return nil, fmt.Errorf("error filtering applications by name: %w", err)
+		}
+	}
+
+	// Filter applications by name
+	newItems = argoutil.FilterAppSetsByProjects(newItems, q.Projects)
+
+	// Sort found applications by name
+	sort.Slice(newItems, func(i, j int) bool {
+		return newItems[i].Name < newItems[j].Name
+	})
+
+	appsetList := v1alpha1.ApplicationSetList{
+		ListMeta: metav1.ListMeta{
+			ResourceVersion: s.appsetInformer.LastSyncResourceVersion(),
+		},
+		Items: newItems,
+	}
+	return &appsetList, nil
+
+}
+
+func (s *Server) Create(ctx context.Context, q *applicationset.ApplicationSetCreateRequest) (*v1alpha1.ApplicationSet, error) {
+	if q.GetApplicationset() == nil {
+		return nil, fmt.Errorf("error creating applicationset: applicationset is nil in request")
+	}
+	if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplicationSet, rbacpolicy.ActionCreate, apputil.AppSetRBACName(q.GetApplicationset())); err != nil {
+		return nil, err
+	}
+
+	appset := q.GetApplicationset()
+
+	project, err := s.validateAppSet(ctx, appset)
+	if err != nil {
+		return nil, fmt.Errorf("error validating applicationset: %w", err)
+	}
+
+	if err := s.checkCreatePermissions(ctx, appset, project); err != nil {
+		return nil, err
+	}
+
+	s.projectLock.RLock(project.Name)
+	defer s.projectLock.RUnlock(project.Name)
+
+	created, err := s.appclientset.ApplicationsetV1alpha1().ApplicationSets(s.ns).Create(ctx, appset, metav1.CreateOptions{})
+	if err == nil {
+		s.logAppSetEvent(created, ctx, argo.EventReasonResourceCreated, "created applicationset")
+		s.waitSync(created)
+		return created, nil
+	}
+
+	if !apierr.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("error creating applicationset: %w", err)
+	}
+	// act idempotent if existing spec matches new spec
+	existing, err := s.appsetLister.Get(appset.Name)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to check existing applicationset details: %v", err)
+	}
+
+	return nil, fmt.Errorf("applicationset with name %s already exists", existing.Name)
+}
+
+func (s *Server) Update(ctx context.Context, q *applicationset.ApplicationSetUpdateRequest) (*v1alpha1.ApplicationSet, error) {
+
+	if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplicationSet, rbacpolicy.ActionUpdate, apputil.AppSetRBACName(q.GetApplicationset())); err != nil {
+		return nil, err
+	}
+
+	validate := true
+	if q.Validate {
+		validate = q.Validate
+	}
+	return s.validateAndUpdateAppSet(ctx, q.Applicationset, false, validate)
+}
+
+func (s *Server) validateAndUpdateAppSet(ctx context.Context, newApp *v1alpha1.ApplicationSet, merge bool, validate bool) (*v1alpha1.ApplicationSet, error) {
+	s.projectLock.RLock(newApp.Spec.Template.Spec.GetProject())
```

Comment:
In my understanding, if the user has included a template parameter in the project field, the ApplicationSet should be rejected that is, the values specified in the project field should be immutable so that the API server can perform necessary operations on them.

cc: @jgwest